File: R_workflow/tests/testthat/test-trial-timing.R
# Per-trial event durations: jittered, or read from an item column.
#
# A fixed duration is a property of the design; one that varies from trial to
# trial is a property of the trial, and therefore a variable the analysis needs.
# Both forms are resolved into the stimuli table before anything is generated, so
# the realised milliseconds travel with the stimuli rather than living only inside
# the presentation script.
#
# The jitter draws no random number. It is a uniform integer keyed on the seed,
# the column name, the list, the set and the condition, so the two engines realise
# the same milliseconds and a rerun reproduces them. Naming the column in the key
# is what makes two jittered events in one design draw independently instead of
# sharing a value.
#
# python_workflow/tests/test_trial_timing.py asserts the same values.

schema <- list(seed = 42L)

events <- list(
  list(type = "fixation", content = "+", duration_ms = 500L),
  list(type = "text", content = "{prime}", duration = list(from_column = "soa_ms"), trigger = 20L),
  list(type = "text", content = "{target}", duration_ms = 800L, trigger = "condition"),
  list(type = "blank", duration = list(jitter = c(400L, 800L), as = "isi_ms"))
)

stim <- data.frame(
  set = 1:4,
  condition = c("related", "unrelated", "related", "unrelated"),
  list = c(1L, 1L, 2L, 2L),
  prime = c("cat", "sun", "dog", "car"),
  target = c("pet", "sky", "pet", "van"),
  soa_ms = c(60L, 250L, 1200L, 60L),
  trial = 1:4,
  stringsAsFactors = FALSE
)

test_that("a jittered duration lands in the stimuli table within its range", {
  out <- resolve_trial_timing(stim, list(events = events), schema)
  expect_true("isi_ms" %in% names(out))
  expect_true(all(out$isi_ms >= 400 & out$isi_ms <= 800))
  # Deterministic: the same inputs give the same milliseconds every time.
  expect_equal(out$isi_ms, resolve_trial_timing(stim, list(events = events), schema)$isi_ms)
})

test_that("the realised values match the Python engine", {
  out <- resolve_trial_timing(stim, list(events = events), schema)
  expect_equal(out$isi_ms, c(537L, 583L, 764L, 594L))
})

test_that("the seed changes the draw", {
  a <- resolve_trial_timing(stim, list(events = events), list(seed = 42L))$isi_ms
  b <- resolve_trial_timing(stim, list(events = events), list(seed = 43L))$isi_ms
  expect_false(identical(a, b))
})

test_that("two jittered events draw independently", {
  ev <- list(
    list(type = "blank", duration = list(jitter = c(0L, 1000L), as = "gap_a_ms")),
    list(type = "blank", duration = list(jitter = c(0L, 1000L), as = "gap_b_ms"))
  )
  out <- resolve_trial_timing(stim, list(events = ev), schema)
  # Sharing a key would make these identical; the column name is in the key.
  expect_false(identical(out$gap_a_ms, out$gap_b_ms))
  expect_equal(out$gap_a_ms, c(76L, 51L, 593L, 690L))
  expect_equal(out$gap_b_ms, c(460L, 499L, 442L, 125L))
})

test_that("a column duration is referenced rather than baked in", {
  rendered <- render_events(events, list(), 60)
  expect_equal(rendered[[1]]$ms, 500L)              # fixed, resolved at build time
  expect_equal(rendered[[2]]$ms_column, "soa_ms")   # per trial, resolved at run time
  # [[ ]], not $: `$` would partial-match `ms_column` and return it.
  expect_null(rendered[[2]][["ms"]])
  expect_equal(rendered[[4]]$ms_column, "isi_ms")
})

test_that("the loop table carries every per-trial duration column", {
  out <- resolve_trial_timing(stim, list(events = events), schema)
  tab <- lexsync:::loop_table(out, events)
  expect_true(all(c("soa_ms", "isi_ms") %in% names(tab)))
})

test_that("a missing source column is an error, not a silent default", {
  ev <- list(list(type = "blank", duration = list(from_column = "not_there")))
  expect_error(resolve_trial_timing(stim, list(events = ev), schema),
               "which the items do not have")
})

test_that("a malformed duration block is rejected", {
  expect_error(render_events(list(list(type = "blank", duration = list(jitter = 400L))), list(), 60),
               "two-element range")
  expect_error(render_events(list(list(type = "blank", duration = list(nonsense = 1L))), list(), 60),
               "from_column' or 'jitter'")
})

test_that("an unnamed jitter column falls back to the event index", {
  ev <- list(type = "blank", duration = list(jitter = c(10L, 20L)))
  expect_equal(lexsync:::.duration_spec(ev, 2L)$column, "event2_ms")
})

test_that("key parts render integral doubles the way Python does", {
  # The hazard this guards: a pandas column promoted to float64 by one missing
  # value would otherwise key on "3.0" in Python and "3" in R.
  expect_equal(lexsync:::.key_part(3L), "3")
  expect_equal(lexsync:::.key_part(3.0), "3")
  expect_equal(lexsync:::.key_part(-7.0), "-7")
  expect_equal(lexsync:::.key_part("related"), "related")
  expect_equal(lexsync:::.key_part(2.5), "2.5")
})

File: python_workflow/tests/test_trial_timing.py
"""Per-trial event durations: jittered, or read from an item column.

A fixed duration is a property of the design; one that varies from trial to trial
is a property of the trial, and therefore a variable the analysis needs. Both
forms are resolved here into the stimuli table before anything is generated, so
the realised milliseconds travel with the stimuli rather than living only inside
the presentation script.

The jitter draws no random number. It is a uniform integer keyed on the seed, the
column name, the list, the set and the condition, so the two engines realise the
same milliseconds and a rerun reproduces them. Naming the column in the key is
what makes two jittered events in one design draw independently instead of
sharing a value.

R_workflow/tests/testthat/test-trial-timing.R asserts the same values.
"""
import pandas as pd
import pytest

from lexsync.scripting import (_duration_spec, _key_part, loop_table, render_events,
                               resolve_trial_timing)

SCHEMA = {"seed": 42}

EVENTS = [
    {"type": "fixation", "content": "+", "duration_ms": 500},
    {"type": "text", "content": "{prime}", "duration": {"from_column": "soa_ms"}, "trigger": 20},
    {"type": "text", "content": "{target}", "duration_ms": 800, "trigger": "condition"},
    {"type": "blank", "duration": {"jitter": [400, 800], "as": "isi_ms"}},
]

STIM = pd.DataFrame({
    "set": [1, 2, 3, 4],
    "condition": ["related", "unrelated", "related", "unrelated"],
    "list": [1, 1, 2, 2],
    "prime": ["cat", "sun", "dog", "car"],
    "target": ["pet", "sky", "pet", "van"],
    "soa_ms": [60, 250, 1200, 60],
    "trial": [1, 2, 3, 4],
})


def test_a_jittered_duration_lands_in_the_stimuli_table_within_its_range():
    out = resolve_trial_timing(STIM, {"events": EVENTS}, SCHEMA)
    assert "isi_ms" in out.columns
    assert out["isi_ms"].between(400, 800).all()
    # Deterministic: the same inputs give the same milliseconds every time.
    assert list(out["isi_ms"]) == list(resolve_trial_timing(STIM, {"events": EVENTS}, SCHEMA)["isi_ms"])


def test_the_realised_values_match_the_r_engine():
    out = resolve_trial_timing(STIM, {"events": EVENTS}, SCHEMA)
    assert list(out["isi_ms"]) == [537, 583, 764, 594]


def test_the_seed_changes_the_draw_and_nothing_else_does():
    a = resolve_trial_timing(STIM, {"events": EVENTS}, {"seed": 42})["isi_ms"]
    b = resolve_trial_timing(STIM, {"events": EVENTS}, {"seed": 43})["isi_ms"]
    assert list(a) != list(b)


def test_two_jittered_events_draw_independently():
    events = [
        {"type": "blank", "duration": {"jitter": [0, 1000], "as": "gap_a_ms"}},
        {"type": "blank", "duration": {"jitter": [0, 1000], "as": "gap_b_ms"}},
    ]
    out = resolve_trial_timing(STIM, {"events": events}, SCHEMA)
    # Sharing a key would make these identical; the column name is in the key.
    assert list(out["gap_a_ms"]) != list(out["gap_b_ms"])


def test_a_column_duration_is_referenced_rather_than_baked_in():
    rendered = render_events(EVENTS, {}, 60)
    assert rendered[0]["ms"] == 500                 # fixed, resolved at build time
    assert rendered[1]["ms_column"] == "soa_ms"     # per trial, resolved at run time
    assert "ms" not in rendered[1]
    assert rendered[3]["ms_column"] == "isi_ms"


def test_the_loop_table_carries_every_per_trial_duration_column():
    out = resolve_trial_timing(STIM, {"events": EVENTS}, SCHEMA)
    tab = loop_table(out, EVENTS)
    assert "soa_ms" in tab.columns and "isi_ms" in tab.columns


def test_a_missing_source_column_is_an_error_not_a_silent_default():
    events = [{"type": "blank", "duration": {"from_column": "not_there"}}]
    with pytest.raises(ValueError, match="which the items do not have"):
        resolve_trial_timing(STIM, {"events": events}, SCHEMA)


def test_a_malformed_duration_block_is_rejected():
    with pytest.raises(ValueError, match="two-element range"):
        render_events([{"type": "blank", "duration": {"jitter": [400]}}], {}, 60)
    with pytest.raises(ValueError, match="from_column' or 'jitter'"):
        render_events([{"type": "blank", "duration": {"nonsense": 1}}], {}, 60)


def test_an_unnamed_jitter_column_falls_back_to_the_event_index():
    events = [{"type": "fixation", "content": "+"}, {"type": "blank",
              "duration": {"jitter": [10, 20]}}]
    assert _duration_spec(events[1], 2)["column"] == "event2_ms"


def test_key_parts_render_integral_floats_like_r_does():
    # The hazard this guards: a pandas column promoted to float64 by one missing
    # value would otherwise key on "3.0" in Python and "3" in R.
    assert _key_part(3) == "3"
    assert _key_part(3.0) == "3"
    assert _key_part(-7.0) == "-7"
    assert _key_part("related") == "related"
    assert _key_part(2.5) == "2.5"


def test_a_hash_key_component_that_cannot_be_rendered_identically_is_refused():
    """Measured divergence before this: a missing value rendered "nan" here and "NA" in
    R, True/False against TRUE/FALSE, inf against Inf. A blank `condition` cell is a
    routine data error that neither reader rejects, and it produced a DIFFERENT trial
    order in each engine -- reproducibly, with nothing to signal it."""
    import pytest
    from lexsync.io_utils import _key_part
    for bad in (None, float("nan")):
        with pytest.raises(ValueError, match="missing"):
            _key_part(bad)
    with pytest.raises(ValueError, match="not finite"):
        _key_part(float("inf"))
    # Booleans are legitimate, so they get a pinned spelling rather than a refusal.
    # test-trial-timing.R asserts the same two strings.
    assert _key_part(True) == "TRUE"
    assert _key_part(False) == "FALSE"
    # A value past the integer range would make R's as.integer() return NA, which would
    # put an empty component into the key.
    assert _key_part(3e9) == "3000000000"
